### DISCUSSION TITLE
BasicPublish buffers

### DISCUSSION THREAD

**Discussion by DwayneBull** [rabbitmq/rabbitmq-dotnet-client#1135] [Q&A]
Is it safe to resuse the ReadOnlyMemory<byte> after calling BasicPublish?
We are trying to pool memory but currently a new buffer is created for each publish as I don't know when it is safe to return them to the pool.

**Comment by lukebakken** [comment]
[`ReadOnlyMemory`](https://docs.microsoft.com/en-us/dotnet/standard/memory-and-spans/) is immutable, so I'm not sure how you would reuse the instance.

**Reply by DwayneBull** [reply]
We have a pool of buffers we use to write data to then publish, we don't want to reuse the struct.
For example:

var buffer = GetPooledBuffer();// This returns something wrapping a byte[] that does serialisation.
buffer.Write(message);// Something that writes some data to the buffer. Message is custom data, e.g. json/bytes or something.
// Raw is accessing the byte[]
// WriteCursor is the length of the buffer we have written to;
var data = new ReadOnlyMemory<byte>(buffer.Raw, 0, buffer.WriteCursor);
// Send to rabbit.
sender.BasicPublish(exchange, routingKey, properties, data);
// Repool the buffer?
// This is where we have the issue, is rabbit still using the underlaying byte[]?

**Reply by lukebakken** [reply]
Here is the code that takes the `ReadOnlyMemory` representing the body and serializes it: [link](https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/main/projects/RabbitMQ.Client/client/impl/Frame.cs#L166-L202)

Notice that the body is copied to a buffer retrieved here: [link](https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/main/projects/RabbitMQ.Client/client/impl/Frame.cs#L176)

> // This is where we have the issue, is rabbit still using the underlaying byte[]?

No, it is not.
